Match numpy sign semantics

prutschman-iv opened the pull request:
Here is my first pass at modifying the `sign` function to match `numpy`'s NaN semantics. It includes a test case using `check_unary_inf_nan`. This is to fix  #8327

I've confirmed that the tests pass on v12, and have rebased to main, but am currently having build difficulties on main, so this is a draft for now.

The non-complex logic is based directly on numpy's logic. The static_casts are there to avoid a compiler error about ambiguous int casting for the ternary operator when the template is used with integer types.

prutschman-iv committed:
Match numpy nan semantics for sign function

prutschman-iv committed:
Add test case for nan semantics to sign function

ev-br commented:
Note that numpy 2.0 redefines the complex sign to return `z/|z|`.

https://github.com/cupy/cupy/pull/8314/commits/cdf5ca83100adace0da42ba008312ea3f392b85f

prutschman-iv committed:
Restore more efficient computation

Once guarded by the isnan test, the (x > 0) - (x < 0) has the correct
semantics, and produces tighter PTX code.

prutschman-iv commented:
I have verified the changes with the main branch. I believe this is ready for merge.

seberg commented:
Sorry that this never moved forward, but considering that it is outdated (with the complex change) and that I got a diff that seems to compile tighter code (even if this is normally memory bound), I'll close it and make a new PR.

EDIT: I'll fetch the test commit for the heck of it!  OK, nvm... that test doesn't work, because the `inf` special cases, etc. so...

leofang commented:
(Continued in #10034.)
